contracts/VestingEscrow.vy

```diff
@@ -0,0 +1,314 @@
+    self.start_time = start_time
+    self.end_time = end_time
+    self.cliff_length = cliff_length
+
+    assert self.token.transferFrom(
+        msg.sender, self, amount
+    ), "could not fund escrow"
+
+    self.total_locked = amount
+    self.recipient = recipient
+    self.disabled_at = end_time  # Set to maximum time
+    self.voting_adapter_addr = voting_adapter_addr
+    log Fund(recipient, amount)
+
+    return True
+
+
+@internal
+@view
+def _total_vested_at(time: uint256 = block.timestamp) -> uint256:
+    start: uint256 = self.start_time
+    end: uint256 = self.end_time
+    locked: uint256 = self.total_locked
+    if time < start + self.cliff_length:
+        return 0
+    return min(locked * (time - start) / (end - start), locked)
+
+
+@internal
+@view
+def _unclaimed(time: uint256 = block.timestamp) -> uint256:
+    return self._total_vested_at(time) - self.total_claimed
+
+
+@external
+@view
+def unclaimed() -> uint256:
+    """
+    @notice Get the number of unclaimed, vested tokens for recipient
+    """
+    # NOTE: if `revoke_unvested` is activated, limit by the activation timestamp
+    return self._unclaimed(min(block.timestamp, self.disabled_at))
+
+
+@internal
+@view
+def _locked(time: uint256 = block.timestamp) -> uint256:
+    if time >= self.disabled_at:
+        return 0
+    return self.total_locked - self._total_vested_at(time)
+
+
+@external
+@view
+def locked() -> uint256:
+    """
+    @notice Get the number of locked tokens for recipient
+    """
+    return self._locked(min(block.timestamp, self.disabled_at))
+
+
+@external
+def claim(
+    beneficiary: address = msg.sender, amount: uint256 = max_value(uint256)
+):
+    """
+    @notice Claim tokens which have vested
+    @param beneficiary Address to transfer claimed tokens to
+    @param amount Amount of tokens to claim
+    """
+    self._check_sender_is_recipient()
+
+    claim_period_end: uint256 = min(block.timestamp, self.disabled_at)
+    claimable: uint256 = min(self._unclaimed(claim_period_end), amount)
+    self.total_claimed += claimable
+
+    assert self.token.transfer(beneficiary, claimable)
+    log Claim(self.recipient, beneficiary, claimable)
+
+
+@external
+def revoke_unvested():
+    """
+    @notice Disable further flow of tokens and revoke the unvested part to owner
+    """
+    self._check_sender_is_owner_or_manager()
+    # NOTE: Revoking more than once is futile
+
+    revokable: uint256 = self._locked()
+    self.disabled_at = block.timestamp
+
+    assert self.token.transfer(self.owner, revokable)
+    log RevokeUnvested(self.recipient, revokable)
+
+
+@external
+def change_owner(owner: address):
+    """
+    @notice Change contract owner.
+    @param owner Address of the new owner. Must be non-zero.
+    """
+    self._check_sender_is_owner()
+    assert owner != empty(address), "zero owner address"
+
+    self.owner = owner
+    log OwnerChanged(owner)
+
+
+@external
+def change_manager(manager: address):
```

Comment:
there's no tests on manager changing
also it should be possible change manager to zero address?